vortex-array/src/array/primitive/compute/filter.rs

```diff
@@ -1,60 +1,63 @@
-use vortex_dtype::{match_each_native_ptype, NativePType};
+use vortex_dtype::match_each_native_ptype;
 use vortex_error::VortexResult;
 
 use crate::array::primitive::PrimitiveArray;
-use crate::compute::{FilterFn, FilterMask};
+use crate::compute::{FilterFn, FilterIter, FilterMask};
 use crate::variants::PrimitiveArrayTrait;
 use crate::{ArrayData, IntoArrayData};
 
 impl FilterFn for PrimitiveArray {
-    fn filter(&self, mask: &FilterMask) -> VortexResult<ArrayData> {
-        filter_select_primitive(self, mask).map(|a| a.into_array())
+    fn filter(&self, mask: FilterMask) -> VortexResult<ArrayData> {
+        let validity = self.validity().filter(&mask)?;
+        match_each_native_ptype!(self.ptype(), |$T| {
+            let mut values = Vec::with_capacity(mask.true_count());
+            match mask.iter()? {
+                FilterIter::Indices(indices) => filter_primitive_indices(&mut values, self.maybe_null_slice::<$T>(), indices.iter().copied()),
+                FilterIter::IndicesIter(iter) => filter_primitive_indices(&mut values, self.maybe_null_slice::<$T>(), iter),
+                FilterIter::Slices(slices) => filter_primitive_slices(&mut values, self.maybe_null_slice::<$T>(), slices.iter().copied()),
+                FilterIter::SlicesIter(iter) => filter_primitive_slices(&mut values, self.maybe_null_slice::<$T>(), iter),
+            }
+            Ok(PrimitiveArray::from_vec(values, validity).into_array())
+        })
     }
 }
 
-fn filter_select_primitive(
-    arr: &PrimitiveArray,
-    mask: &FilterMask,
-) -> VortexResult<PrimitiveArray> {
-    let validity = arr.validity().filter(mask)?;
-    let selection_count = mask.true_count();
-    match_each_native_ptype!(arr.ptype(), |$T| {
-        let slice = arr.maybe_null_slice::<$T>();
-        Ok(PrimitiveArray::from_vec(filter_primitive_slice(slice, mask, selection_count)?, validity))
-    })
+fn filter_primitive_indices<T: Copy>(
+    output: &mut Vec<T>,
```

Comment:
nit: it's VERY weird to put mutable output as first arg, especially when second arg is an input of the approx the same type... 

When I've seen this pattern used previously in both C++ & Rust, it's conventional to put the mutable output args at the end